src/modules/quick_api_module.f90

```diff
@@ -0,0 +1,688 @@
+  quick_api%coords         = coords
+  if(quick_api%nxt_ptchg>0) then 
+    quick_api%xt_chg_crd = xt_chg_crd
+    quick_api%ptchg_grad = ptchg_grad
+  endif
+
+  call run_quick(quick_api)
+
+  ! send back total energy, gradients and point charge gradients
+  energy     = quick_api%tot_ene
+  gradients     = quick_api%gradient
+
+  if(quick_api%nxt_ptchg>0) ptchg_grad = quick_api%ptchg_grad
+
+end subroutine get_quick_energy_gradients
+
+
+! runs quick, partially resembles quick main program
+subroutine run_quick(self)
+
+  use quick_timer_module
+  use quick_method_module, only : quick_method
+  use quick_files_module
+  use quick_calculated_module, only : quick_qm_struct
+  use quick_gridpoints_module, only : quick_dft_grid, deform_dft_grid
+#ifdef MPIV
+  use quick_mpi_module
+#endif
+
+  implicit none
+
+  type(quick_api_type), intent(inout) :: self
+  integer :: ierr, i, j, k
+  logical :: failed = .false.
+
+  ! print step into quick output file
+  call print_step(self)  
+
+  ! if dft is requested, make sure to delete dft grid variables from previous 
+  ! the md step before proceeding
+  if(( self%step .gt. 1 ) .and. quick_method%DFT) call deform_dft_grid(quick_dft_grid)
+
+  ! set molecular information into quick_molspec
+  call set_quick_molspecs(quick_api)
+
+  ! start the timer for initial guess
+  call cpu_time(timer_begin%TIniGuess)
+
+  ! we will reuse density matrix for steps above 1. For the 1st step, we should
+  ! read basis file and run SAD guess. 
+  if(self%firstStep .and. self%reuse_dmx) then
+
+    ! perform the initial guess
+    if (quick_method%SAD) call getMolSad()  
+
+    ! assign basis functions 
+    call getMol()
+
+    self%firstStep = .false.
+
+  endif
+
+  ! pre-calculate 2 index coefficients and schwarz cutoff criteria
+  if(.not.quick_method%opt) then
+    call g2eshell
+    call schwarzoff
+  endif
+
+#if defined CUDA || defined CUDA_MPIV    
+  ! upload molecular and basis information to gpu
+  call gpu_upload_molspecs()
+#endif
+
+  ! stop the timer for initial guess
+  call cpu_time(timer_end%TIniGuess)
+
+  ! compute energy
+  if ( .not. quick_method%opt .and. .not. quick_method%grad) then
+    call getEnergy(failed)
+  endif
+  
+  ! compute gradients
+  if ( .not. quick_method%opt .and. quick_method%grad) call gradient(failed)
+
+  ! run optimization
+  if (quick_method%opt)  call optimize(failed)
+
+#if defined CUDA || defined CUDA_MPIV
+      if (quick_method%bCUDA) then
+        call gpu_cleanup()
+      endif
+#endif
+
+  if (failed) call quick_exit(iOutFile,1)
+
+#ifdef MPIV
+  if(master) then
+#endif
+
+  ! calculate charges
+  if (quick_method%dipole) call dipole
+
+! save the results in quick_api struct
+  self%tot_ene = quick_qm_struct%Etot
+
+! convert gradients into gradients and save in quick_api struct &
+! save point charge gradients too. Note that quick_qm_struct saves
+! both gradients in vector formats and we should organize them
```

Comment:
In principle the layout in memory should be the same since Fortran stores in column major format, i.e. we have x1, y1, z1, x2, y2, z2, ... both in quick_api struct and quick_qm_struct. But being explicit is probably better.

Reply:
I agree. I also corrected the comment.